Implement stacked layout

RyanDwyer opened the pull request:
You know the drill. It's mostly a copy/paste from tabbed.

At some point (probably after floating) I will try to combine some of the rendering parts to minimise the amount of duplicate/similar code.

emersion commented:
>I will try to combine some of the rendering parts to minimise the amount of duplicate/similar code.

Is it possible to do that now?

emersion commented on sway/desktop/output.c:
```diff
@@ -790,12 +790,177 @@ static void render_container_tabbed(struct sway_output *output,
+	float color[4];
+
+	if (view->border != B_NONE) {
+		if (view->border_left) {
```
Maybe we could add functions to render borders that could be reused?

RyanDwyer commented:
>> I will try to combine some of the rendering parts to minimise the amount of duplicate/similar code.

> Is it possible to do that now?

Done. Removed more lines of code than I added :D

Also, I checked i3's implementation and made it pixel perfect. Marks now show in titlebars for tabbed and stacked containers too.

RedSoxFan commented:
The only issue that I've noticed so far is that you cannot move a container out of the stack in either vertical direction. It will only move up and down the stack and stop when hitting the bounds. It appears tabs have the same issue in the horizontal direction. 

If you have multiple outputs, moving a container out of the stack/tabs in the direction of the other output does move it out, but to the other output. To match i3, you should be able to move containers out of stacks/tabs and create a split on a single output (then if you move again, it will move to another outputs, if possible).

SirCmpwn commented:
>The only issue that I've noticed so far is that you cannot move a container out of the stack in either vertical direction. It will only move up and down the stack and stop when hitting the bounds. It appears tabs have the same issue in the horizontal direction.

Since this affects tabs too, you should file an issue and we'll address it separately.

emersion commented on sway/desktop/output.c:
```diff
@@ -317,108 +306,159 @@ static void premultiply_alpha(float color[4], float opacity) {
+		}
+	}
+	box.x = x + left_offset;
+	box.y = y + config->font_height + 7;
```
Is it possible to add the 3px padding and the 1px border as constants to know why there are 7, 6, 4 and 3 floating around?

emersion commented:
Code LGTM, nice work!

Rendering looks good, but pointer input to select the stack titlebar doesn't work (works in tabbed mode).

ascent12 commented on include/sway/tree/container.h:
```diff
@@ -212,4 +218,9 @@ void container_calculate_title_height(struct sway_container *container);
+/**
+ * Return the height of a regular title bar.
+ */
+size_t container_titlebar_height();
```
Very small point, but you should put an explicit `(void)` as the arguments for a prototype.
An empty argument list for prototypes has surprising results in C.

RyanDwyer replied:
Ah yeah, you or someone else has mentioned that to me before. I'll change it.

ascent12 replied:
I may as well give the reason now, in case anyone stumbles on this comment again.
An empty argument list in a prototype is actually declaring a function with _unspecified_ arguments, not no arguments.
So if you tried to call it like `container_titlebar_height(100);`, you wouldn't get a warning.

SirCmpwn commented:
Great work! Thanks!
